 jni: consider /data/app to the fd whitelist if Xposed is detected

Latest security update has added whitelisting routine to the Zygote.

Since Xposed now reads from /data/app,
it's necessary to add /data/app to the whitelist.

Dynamically do this if XposedBridge.jar is detected.

diff --git a/core/jni/fd_utils-inl.h b/core/jni/fd_utils-inl.h
index 3c1c0bc..0d9e0e0 100644
--- a/core/jni/fd_utils-inl.h
+++ b/core/jni/fd_utils-inl.h
@@ -306,6 +306,16 @@ class FileDescriptorInfo {
       return true;
     }
 
+    if (access("/system/framework/XposedBridge.jar", F_OK ) != -1) {
+      // Xposed-powered Zygote might read from extensions other than .apk
+      // so skip extension check
+      ALOGW("Xposed detected, loosening up Zygote fd check!");
+      static const std::string kDataAppPrefix = "/data/app/";
+      if (path.compare(0, kDataAppPrefix.size(), kDataAppPrefix) == 0) {
+        return true;
+      }
+    }
+
     return false;
   }
 
